xdem/coreg.py

```diff
@@ -1063,3 +1081,620 @@ def coregister(reference_raster: Union[str, gu.georaster.Raster], to_be_aligned_
+        :param rejection_scale: The threshold (std * rejection_scale) to consider points as outliers.
+        :param num_levels: Number of octree levels to consider. A higher number is faster but may be more inaccurate.
+        """
+        if not _has_cv2:
+            raise ValueError("Optional dependency needed. Install 'opencv'")
+        self.max_iterations = max_iterations
+        self.tolerance = tolerance
+        self.rejection_scale = rejection_scale
+        self.num_levels = num_levels
+        self._meta: dict[str, Any] = {}
+
+    def _fit_func(self, ref_dem: np.ndarray, tba_dem: np.ndarray, transform: Optional[rio.transform.Affine],
+                  weights: Optional[np.ndarray]):
+        """Estimate the rigid transform from tba_dem to ref_dem."""
+        if weights is not None:
+            warnings.warn("ICP was given weights, but does not support it.")
+
+        bounds, resolution = _transform_to_bounds_and_res(ref_dem, transform)
+        points: dict[str, np.ndarray] = {}
+        # Generate the x and y coordinates for the reference_dem
+        x_coords, y_coords = np.meshgrid(
+            np.linspace(bounds.left + resolution / 2, bounds.right - resolution / 2, num=ref_dem.shape[1]),
+            np.linspace(bounds.bottom + resolution / 2, bounds.top - resolution / 2, num=ref_dem.shape[0])[::-1]
+        )
+        # Subtract by the bounding coordinates to avoid float32 rounding errors.
+        x_coords -= bounds.left
+        y_coords -= bounds.bottom
+        for key, dem in zip(["ref", "tba"], [ref_dem, tba_dem]):
+
+            gradient_x, gradient_y = np.gradient(dem)
+
+            normal_east = np.sin(np.arctan(gradient_y / resolution)) * -1
+            normal_north = np.sin(np.arctan(gradient_x / resolution))
+            normal_up = 1 - np.linalg.norm([normal_east, normal_north], axis=0)
+
+            valid_mask = ~np.isnan(dem) & ~np.isnan(normal_east) & ~np.isnan(normal_north)
+
+            point_cloud = np.dstack([
+                x_coords[valid_mask],
+                y_coords[valid_mask],
+                dem[valid_mask],
+                normal_east[valid_mask],
+                normal_north[valid_mask],
+                normal_up[valid_mask]
+            ]).squeeze()
+
+            points[key] = point_cloud[~np.any(np.isnan(point_cloud), axis=1)].astype("float32")
+
+        icp = cv2.ppf_match_3d_ICP(self.max_iterations, self.tolerance, self.rejection_scale, self.num_levels)
+        _, residual, matrix = icp.registerModelToScene(points["tba"], points["ref"])
+
+        assert residual < 1000, f"ICP coregistration failed: residual={residual}, threshold: 1000"
+
+        self._meta["matrix"] = matrix
+
+    def _apply_func(self, dem: np.ndarray, transform: rio.transform.Affine) -> np.ndarray:
+        """Apply the coregistration matrix to a DEM."""
+        bounds, resolution = _transform_to_bounds_and_res(dem, transform)
```

Comment:
This call seem to almost always be followed by meshgrid to generate a coordinate grid. Maybe have a specific function for this? I think in case for example we implement a subsampling in the future, this would have to be replaced only once.
The only cases where _transform_to_bounds is used is, I believe, to get the resolution, which is actually contained in the transform itself.

Reply:
I'm pretty sure the resolution is not contained in the transform. [See the rio docs](https://rasterio.readthedocs.io/en/latest/api/rasterio.transform.html#rasterio.transform.array_bounds) where the height and width have to be supplied each time a raster-specific operation is made. I believe the Affine transform only contains information of the upper left corner, the scale, and the rotation, but I might be wrong.

Anyway, I cleaned up the code a bit and made a function for x_coords and y_coords  creation, so we can more easily change this in the future if we want to (56a4d1b)

Reply:
The resolution is in transform[0] and transform[4]. But I agree, to get the bounds, you need the shape. 
My reply contained a typo, what I meant is that in at least one occasion, _transform_to_bounds_and_res was not followed by meshgrid, and in that case, only the resolution was needed, not the bounds, making the call to this function useless.